pkg/cli/init_project.go

```diff
@@ -0,0 +1,135 @@
+)
+
+func (c cli) newInitProjectCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "init",
+		Short: "Initialize a new project",
+		Long: `Initialize a new project.
+
+For further help about a specific project version, set --project-version.
+`,
+		Run:     func(cmd *cobra.Command, args []string) {},
+		Example: c.getInitHelpExamples(),
+	}
+
+	// Register --project-version on the dynamically created command
+	// so that it shows up in help and does not cause a parse error.
+	cmd.Flags().String("project-version", c.defProjVersion,
+		fmt.Sprintf("project version, (%s)", strings.Join(c.getAvailableProjectVersions(), ", ")))
+
+	// Pre-parse the project version and help flags so that we can
+	// dynamically bind to a plugin's init implementation (or not).
+	projectVersion, isHelpOnly := c.getBaseFlags()
+
+	// If only the help flag was set, return the command as is.
+	if isHelpOnly {
+		return cmd
+	}
+
+	// Lookup the plugin for projectVersion and bind it to the command.
+	c.bindInit(cmd, projectVersion)
+	return cmd
+}
+
+func (c cli) getInitHelpExamples() string {
+	vs := c.getAvailableProjectVersions()
+	var sb strings.Builder
+
+	for _, v := range vs {
+		rendered := fmt.Sprintf(`  # Help for initializing a project with version %s
+  %s init --project-version=%s -h
+
+`, v, c.commandName, v)
+		sb.WriteString(rendered)
+	}
+	return strings.TrimSuffix(sb.String(), "\n\n")
+}
+
+func (c cli) getAvailableProjectVersions() (projectVersions []string) {
+	for _, ps := range c.plugins {
+		for _, p := range ps {
+			// Only return project versions from non-deprecated plugins
+			if _, ok := p.(plugin.Deprecated); ok {
+				continue
+			}
+			if _, ok := p.(plugin.Init); !ok {
+				continue
+			}
+			projectVersions = append(projectVersions, strconv.Quote(p.Version()))
+		}
+	}
+	return projectVersions
+}
+
+func (c cli) bindInit(cmd *cobra.Command, projectVersion string) {
+	ps, ok := c.plugins[projectVersion]
+	if !ok {
+		log.Fatal(fmt.Errorf("unknown project version %q", projectVersion))
+	}
+	var getter plugin.InitPluginGetter
+	var hasGetter bool
+	for _, p := range ps {
+		tmpGetter, isGetter := p.(plugin.InitPluginGetter)
+		if isGetter {
+			if hasGetter {
+				log.Fatalf("duplicate initialization plugins for project version %q: %s, %s",
+					projectVersion, getter.Name(), p.Name())
+			}
+			hasGetter = true
+			getter = tmpGetter
+		}
+	}
+	if !hasGetter {
+		log.Fatalf("project version %q does not support a project initialization plugin",
+			projectVersion)
+	}
+
+	ip := getter.GetInitPlugin()
+	ip.BindFlags(cmd.Flags())
+	// TODO: inject defaults.
+	ctx := plugin.Context{
+		CommandName: c.commandName,
+	}
+	ip.UpdateContext(&ctx)
+	ip.SetVersion(projectVersion)
+	cmd.Long = ctx.Description
+	cmd.Example = ctx.Examples
+	cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		if internal.Configured() {
+			return fmt.Errorf("failed to initialize project because project is already initialized")
+		}
```

Comment:
Shouldn't this check be left to each init command? Maybe in a future we want to allow `kubebuilder init` on previously configured projects to change some configuration setting or something.

Without this part `RunE` will be very similar for the init and the create api commands, and thus you could make a helper function.

Reply:
NOTE: this applies to `pkg/cli/api.go`, `pkg/cli/init.go` and `pkg/cli/webhook.go`.

Checking if the project is configured is the only difference between these three, which means that if we remove it we can make a helper that creates the RunE function for us similar to `errCmdFunc`. This check is already being done by the `LoadConfig` step of `RunOptions` (`internal/cmdutil/cmdutil.go:28`), so it is redundant.